imageBuilder.go

```diff
@@ -9,32 +9,65 @@ import (
 
 var publishImage = flag.Bool("publishImage", false, "Push image to docker hub")
 var codeRunnerConfig = flag.String("codeRunnerConfig", "code-runner.yaml", "Code Runner configuration filename")
+var assignmentEnvConfig = flag.String("assignmentEnvConfig", "assignment-env.yaml", "Assignment Environment configuration filename")
 var dockerAuthConfig = flag.String("dockerAuthConfig", "docker-auth.yaml", "Docker hub authentication configuration filename")
 var dockerfileName = flag.String("dockerfile", "Dockerfile", "Dockerfile name")
 
 func main() {
+
 	flag.Parse()
 
 	log.Println("Creating Dockerfile...")
 
+	imgBuilder := builder.NewImageBuilder(*dockerAuthConfig, *dockerfileName)
+
+	err := generateCodeRunnerImage(imgBuilder)
+	if err == nil {
+		_ = generateAssignmentEnvImage(imgBuilder)
+	}
+
+}
+
+// Generate a dockerfile for code runner server, build its image and push it ot docker hub.
+func generateCodeRunnerImage(imgBuilder *builder.ImageBuilder) error {
```

Comment:
Why do we need this extra function?
Seems like the function name is misleading as we already have a `BuildImage(...)`

Reply:
This function is there as of now. Since the code-runner image is being built through code itself. Also, `BuildImage()` is a part of `builder` package. It only builds the image locally. `generateCodeRunnerImage()` is in main wherein the entire process of building and publishing the image is taken care of.

Reply:
> Since the code-runner image is being built through code itself.
Didn't understand what this means.

I still don't see why the image builder cannot do everything. Is not its job to build an image using user provided requirements?
I feel that the ImageBuilder can be designed like this.

```go
type ImageBuilder struct {
    authData *dockerAuthData
    imageTag string // could be anything.
    dockerfileName string
}

type ImageBuilderOption func(*ImageBuilder) error

func NewImageBuilder(options ...ImageBuilderOption) (*ImageBuilder, error) {
    b := &ImageBuilder{}
    for _, opt := range options {
        if err := opt(b); err != nil {
            return errors.Wrap(err, "failed to build ImageBuilder instance")
        }
    }
    return b, nil
}

func WithDockerfileName(filename string) ImageBuilderOption {
    return func(b *ImageBuilder) error {
        // Validate filename and raise error if validation fails.
        b.dockerfileName = filename
    }
}

func WithDockerAuthData(authData *dockerAuthData) ImageBuilderOption {
    return func(b *ImageBuilder) error {
        // Validate authData and raise error if validation fails.
        b.authData = authData
    }
}

func WithImageTag(tag string) ImageBuilderOption {
    return func(b *ImageBuilder) error {
        // Validate tag and raise error if validation fails.
        b.imageTag = tag
    }
}
```

You would then create an instance like this.
```go
imgBuilder, err := NewImageBuilder(
         WithDockerAuthData(authData),
         WithImageTag(getImageTag(config)),
         WithDockerfileName(dockerfileName))
```

This will prevent you from passing unnecessary input to ImageBuilder functions.
You can then just have two functions `BuildImage()` and `BuildAndPushImage()`
that function as the name suggests.

Reply:
Okay. So I have done as suggested.

Reply:
So as per this design, two instances of `ImageBuilder` are created in main. One is for `code-runner` and second is for `assignment-env`. Accordingly the `Build()` and `Publish()` functions of respective instances are called.

Reply:
Why do we need two instances of image builder?